src/vm/database/structures.rs

```diff
@@ -93,3 +93,114 @@ clarity_serializable!(i128);
 clarity_serializable!(u128);
 clarity_serializable!(u64);
 clarity_serializable!(Contract);
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct STXBalance {
+    pub amount_unlocked: u128,
+    pub amount_locked: u128,
+    pub unlock_height: u64,
+}
+clarity_serializable!(STXBalance);
+
+#[derive(Debug)]
+pub enum STXBalanceError {
+    Overflow,
+    Underflow,
+    LockActive,
+}
+
+type Result<T> = std::result::Result<T, STXBalanceError>;
+
+impl STXBalance {
+
+    pub fn zero() -> STXBalance {
+        STXBalance {
+            amount_unlocked: 0,
+            amount_locked: 0,
+            unlock_height: 0,
+        }
+    }
+
+    pub fn get_available_balance_at_block(&self, block_height: u64) -> u128 {
+        match self.has_locked_tokens_unlockable(block_height) {
+            true => self.get_total_balance(),
+            false => self.amount_unlocked
+        }
+    }
+
+    pub fn lock_tokens(&mut self, amount_to_lock: u128, unlock_height: u64, current_height: u64) -> Result<()> {
```

Comment:
So, the reason the lock/unlock logic was all inlined with database loads/stores in the previous PR was because I didn't want there to be any unnecessary representations (or additional mutable representations) of an account's token lock state.  Before, there were only two representations:  the _read-write_ database, and a _read-only_ copy of the account as it was loaded.  This was intentional, because it made it _impossible_ to lock tokens _without also_ locking them in the database.

This method, and this representation of STX balance and lock state, makes me deeply uncomfortable.  It's possible to write buggy code _that will compile_ that calls this method and one of the unlock methods multiple times, and in doing so, create a representation of an account's balance within the `STXBalance` instance that is not consistent with the blockchain transactions' token movements.

I think any methods here that take `&mut self` should be removed, and the necessary logic they contain re-inlined with the database operations that load/store account balances and lock state.

Reply:
> buggy code that will compile that calls this method and one of the unlock methods multiple times, and in doing so, create a representation of an account's balance within the STXBalance instance that is not consistent with the blockchain transactions' token movements

Do you think you can write a snippet of code that would illustrates and give more context? That would be helpful. In my mind, if a developer wants an immutable instance of the balance, they will just declare the var as non mutable, and won't be able to call the mutating functions? 

I think it's in our interest to DRY as much as possible the critical sections of our codebase, and I would qualify the balance logic as critical. 
I felt uncomfortable when I saw that the **logic** around the balance was inlined, repeated, and coupled with the MARF (which increase the difficulty to test it). That was perfectly fine when the only thing we were doing was `checked_add` and `checked_sub`, but now that the logic around locks is introducing subtleties, it might be time to rely on some helpers (and I feel like it already paid off, since this refactoring helped catching issues that we missed during the review).

Reply:
The invariant that I want to be enforced here is that anytime you alter the lock state for an account, you _also_ do so in the database.  This invariant was enforced by the previous PR, but it is no longer enforced here.

To see why this invariant is important, consider this hypothetical code path with the new methods:

```
let mut account = StacksChainState::get_account(clarity_tx, alice)?;
account.stx_balance.lock_tokens(alice_lockup, alice_unlock_height, cur_block_height)?;

/* At this point, account.stx_balance and the Clarity DB are _inconsistent _. */
/* To be safe, you should _immediately_ save the new lock state. */
/* If you fail to do so, then suppose now that this code subsequently runs: */
let mut spending_account = StacksChainState::get_account(clarity_tx, alice)?;
spending_account.debit(amount, cur_block_height)?;  // MAY SUCCEED, even if _all_ of Alice's tokens were supposed to be locked!
```

With my original PR, the attempt to debit `alice`'s account is guaranteed to fail, because the second call to `get_account` would always read the latest locked state.  This property was enforced by ensuring that the only way to lock tokens was by also altering the database.

With this new PR, it is possible to accidentally write code that succeeds in debiting `alice`'s account, because you forgot to also save the lock state.  While it is possible to avoid the above bug, it's _harder_ to do so, because now each time you call `lock_tokens` (or do some other mutation on `STXBalance`), you have to manually ensure that each such instance is coupled with a corresponding set of calls to mutate the state in the Clarity database.  The point I'm trying to emphasize here is that **this is more dangerous than what we had been doing before**, because it makes correctness dependent on us _humans_ always manually verifying that all mutations are materialized, even months and years later when we're maintaining this code (whereas before, we didn't have to think about this at all).

I'm not opposed to helper methods and DRY'ing code -- in fact, I encourage both.  But not at the expense of correctness (goes without saying), and not at the expense of making the codebase more dangerous to work with.

Reply:
Per the public blockchain call, we can leave this as is for now.  But, I intend to address this later, post Krypton.  https://github.com/blockstack/stacks-blockchain/issues/1878